File: zig/presence/publisher.zig
//! zig/presence -- publisher.
//!
//! LIVELINESS QoS reference app, talking to zzdds's native Zig API directly.
//! See docs/design/presence-reference-app.md at the repo root for the full
//! spec. Demonstrates MANUAL_BY_TOPIC_LIVELINESS_QOS: the writer writes for
//! a while, then deliberately goes quiet (no writes, no asserts) for longer
//! than its own lease_duration -- letting the lease actually expire -- then
//! calls DataWriter.assert_liveliness() explicitly before resuming.
//!
//! Required stdout markers (see the spec doc): "Create topic:", "Create
//! writer for topic:", "Publisher: wrote sequence=", "Publisher: going
//! offline", "Publisher: asserting liveliness and resuming", "Publisher:
//! done." Any failure path prints a line starting "FAIL:" and exits nonzero.

const std = @import("std");
const zzdds = @import("zzdds");
const DDS = @import("zzdds_generated").DDS;
const ZZDDS = @import("zzdds_ext_generated").zzdds;
const presence_gen = @import("presence_gen");

const ONLINE_BEACON_COUNT: i32 = 8;
const BEACON_PERIOD_NS: u64 = 500 * std.time.ns_per_ms;
const LEASE_DURATION_S: i32 = 2;
const OFFLINE_DURATION_NS: u64 = 5 * std.time.ns_per_s; // > LEASE_DURATION_S
const READER_READY_TIMEOUT_NS: i64 = 10 * std.time.ns_per_s;
const DRAIN_TIMEOUT_NS: i64 = 15 * std.time.ns_per_s;
const POLL_PERIOD_NS: u64 = 20 * std.time.ns_per_ms;

// ── Time helpers (std.Io.Clock -- portable across Linux/macOS/Windows) ──────

fn monoNs(io: std.Io) i64 {
    return @intCast(std.Io.Clock.awake.now(io).nanoseconds);
}

fn sleepNs(io: std.Io, ns: u64) void {
    (std.Io.Clock.Duration{ .raw = .{ .nanoseconds = @intCast(ns) }, .clock = .awake }).sleep(io) catch {};
}

// ── Listener state, shared between the network thread (callbacks) and main ──

const State = struct {
    reader_ready: std.atomic.Value(bool) = std.atomic.Value(bool).init(false),
    matched_current_count: std.atomic.Value(i32) = std.atomic.Value(i32).init(0),
    ever_matched: std.atomic.Value(bool) = std.atomic.Value(bool).init(false),
};

fn onReliableReaderReady(state: *State, reader_handle: DDS.InstanceHandle_t, is_ready: bool) void {
    _ = reader_handle;
    if (is_ready) state.reader_ready.store(true, .release);
    std.debug.print("on_reliable_reader_ready() is_ready={}\n", .{is_ready});
}

fn onPublicationMatched(state: *State, dw: DDS.DataWriter, status: DDS.PublicationMatchedStatus) void {
    _ = dw;
    state.matched_current_count.store(status.current_count, .release);
    if (status.current_count > 0) state.ever_matched.store(true, .release);
    std.debug.print("on_publication_matched() current_count={d}\n", .{status.current_count});
}

// ── Argument parsing ─────────────────────────────────────────────────────────

fn parseDomain(process_args: std.process.Args) u32 {
    var it = std.process.Args.Iterator.init(process_args);
    _ = it.skip(); // program name
    while (it.next()) |arg| {
        if (std.mem.eql(u8, arg, "-d") or std.mem.eql(u8, arg, "--domain")) {
            const v = it.next() orelse continue;
            return std.fmt.parseInt(u32, v, 10) catch 0;
        }
    }
    return 0;
}

// ── main ──────────────────────────────────────────────────────────────────────

pub fn main(init: std.process.Init) !void {
    const io = init.io;
    var gpa = std.heap.DebugAllocator(.{}){};
    defer _ = gpa.deinit();
    const alloc = gpa.allocator();

    const domain_id = parseDomain(init.minimal.args);

    var factory = zzdds.createFactory() catch {
        std.debug.print("FAIL: createFactory() failed\n", .{});
        std.process.exit(1);
    };
    defer factory.deinit();
    const dpf = factory.toDDSFactory();

    const dp = dpf.create_participant(domain_id, .{}, null, 0);
    if (dp.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_participant() failed on domain {d}\n", .{domain_id});
        std.process.exit(1);
    }
    defer _ = dpf.delete_participant(dp);

    var ts_alloc = alloc;
    if (!zzdds.registerTypeSupport(dp, "PresenceBeacon", .{
        .ctx = @ptrCast(&ts_alloc),
        .compute_key_hash = presence_gen.PresenceBeacon.computeKeyHashFromCdr,
    })) {
        std.debug.print("FAIL: registerTypeSupport() failed\n", .{});
        std.process.exit(1);
    }

    const topic = dp.create_topic("PresenceBeacon", "PresenceBeacon", .{}, null, 0);
    if (topic.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_topic() failed\n", .{});
        std.process.exit(1);
    }
    std.debug.print("Create topic: PresenceBeacon\n", .{});

    const publisher = dp.create_publisher(.{}, null, 0);
    if (publisher.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_publisher() failed\n", .{});
        std.process.exit(1);
    }

    var dw_qos = DDS.DataWriterQos{};
    dw_qos.reliability.kind = .RELIABLE_RELIABILITY_QOS;
    dw_qos.history.kind = .KEEP_LAST_HISTORY_QOS;
    dw_qos.history.depth = 1;
    dw_qos.liveliness.kind = .MANUAL_BY_TOPIC_LIVELINESS_QOS;
    dw_qos.liveliness.lease_duration = .{ .sec = LEASE_DURATION_S, .nanosec = 0 };

    const dw = publisher.create_datawriter(topic, dw_qos, null, 0);
    if (dw.ptr == zzdds.dcps.NIL_PTR) {
        std.debug.print("FAIL: create_datawriter() failed\n", .{});
        std.process.exit(1);
    }
    std.debug.print("Create writer for topic: PresenceBeacon\n", .{});

    var state = State{};
    const zdw = zzdds.asZzddsDataWriter(dw) orelse {
        std.debug.print("FAIL: asZzddsDataWriter() failed\n", .{});
        std.process.exit(1);
    };
    if (zdw.set_listener_ex(ZZDDS.dataWriterListenerEx(&state, .{
        .on_publication_matched = onPublicationMatched,
        .on_reliable_reader_ready = onReliableReaderReady,
    }), DDS.PUBLICATION_MATCHED_STATUS) != DDS.RETCODE_OK) {
        std.debug.print("FAIL: set_listener_ex failed\n", .{});
        std.process.exit(1);
    }

    const writer = presence_gen.PresenceBeaconDataWriter.init(dw, alloc);

    const ready_deadline = monoNs(io) + READER_READY_TIMEOUT_NS;
    while (!state.reader_ready.load(.acquire)) {
        if (monoNs(io) > ready_deadline) {
            std.debug.print("FAIL: no reliable reader became ready within 10s\n", .{});
            std.process.exit(1);
        }
        sleepNs(io, POLL_PERIOD_NS);
    }

    // ── Online phase ─────────────────────────────────────────────────────
    var seq: i32 = 0;
    while (seq < ONLINE_BEACON_COUNT) : (seq += 1) {
        writer.write(.{ .seq_num = seq }, 0) catch {
            std.debug.print("FAIL: write() failed at sequence={d}\n", .{seq});
            std.process.exit(1);
        };
        std.debug.print("Publisher: wrote sequence={d}\n", .{seq});
        sleepNs(io, BEACON_PERIOD_NS);
    }

    // ── Offline phase -- no writes, no asserts, longer than the lease ──────
    std.debug.print("Publisher: going offline (no writes/asserts for {d}s, lease is {d}s)\n", .{ OFFLINE_DURATION_NS / std.time.ns_per_s, LEASE_DURATION_S });
    sleepNs(io, OFFLINE_DURATION_NS);

    // ── Recovery ─────────────────────────────────────────────────────────
    std.debug.print("Publisher: asserting liveliness and resuming\n", .{});
    if (dw.assert_liveliness() != DDS.RETCODE_OK) {
        std.debug.print("FAIL: assert_liveliness() failed\n", .{});
        std.process.exit(1);
    }

    while (seq < ONLINE_BEACON_COUNT * 2) : (seq += 1) {
        writer.write(.{ .seq_num = seq }, 0) catch {
            std.debug.print("FAIL: write() failed at sequence={d}\n", .{seq});
            std.process.exit(1);
        };
        std.debug.print("Publisher: wrote sequence={d}\n", .{seq});
        sleepNs(io, BEACON_PERIOD_NS);
    }

    // Wait for the subscriber to tear its reader down before exiting.
    const drain_deadline = monoNs(io) + DRAIN_TIMEOUT_NS;
    while (!(state.ever_matched.load(.acquire) and state.matched_current_count.load(.acquire) == 0)) {
        if (monoNs(io) > drain_deadline) {
            std.debug.print("FAIL: subscriber did not disconnect within 15s\n", .{});
            std.process.exit(1);
        }
        sleepNs(io, POLL_PERIOD_NS);
    }

    std.debug.print("Publisher: done.\n", .{});
}

File: c/catchup/src/publisher.c
/*
 * c/catchup -- publisher. Direct C port of zig/catchup/publisher.zig; see
 * docs/design/catchup-reference-app.md at the repo root for the full spec.
 * Writes a historical batch immediately, with no reader matched yet, then
 * -- once a reader does match -- writes a live batch. TRANSIENT_LOCAL
 * durability means zzdds's own writer-side cache (not this app) is what
 * makes the historical batch replayable to a late joiner.
 *
 * Required stdout markers (see the spec doc): "Create topic:", "Create
 * writer for topic:", "Publisher: wrote historical seq_num=", "Publisher:
 * reader matched, writing live batch", "Publisher: wrote live seq_num=",
 * "Publisher: done." Any failure path prints a line starting "FAIL:" and
 * exits nonzero.
 */
#include "catchup_sample.h"
#include "zzdds_c.h"

#include <stdatomic.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define HISTORICAL_COUNT 10
#define LIVE_COUNT 5
#define MATCH_TIMEOUT_MS 15000
#define DRAIN_TIMEOUT_MS 15000
#define POLL_PERIOD_MS 20

typedef struct {
    atomic_bool ever_matched;
    atomic_int matched_current_count;
} PubState;

static void on_publication_matched(DDS_DataWriter writer, const DDS_PublicationMatchedStatus *status, void *listener_data) {
    (void)writer;
    PubState *state = (PubState *)listener_data;
    atomic_store(&state->matched_current_count, status->current_count);
    if (status->current_count > 0) atomic_store(&state->ever_matched, true);
    printf("on_publication_matched() current_count=%d\n", status->current_count);
}

static uint32_t parse_domain(int argc, char **argv) {
    for (int i = 1; i < argc - 1; i++) {
        if (strcmp(argv[i], "-d") == 0 || strcmp(argv[i], "--domain") == 0) {
            return (uint32_t)strtoul(argv[i + 1], NULL, 10);
        }
    }
    return 0;
}

int main(int argc, char **argv) {
    uint32_t domain_id = parse_domain(argc, argv);

    zzdds_DomainParticipantFactory factory = zzdds_create_factory();
    if (zzdds_factory_is_nil(factory)) {
        fprintf(stderr, "FAIL: createFactory() failed\n");
        return 1;
    }
    DDS_DomainParticipantFactory dds_factory = zzdds_DomainParticipantFactory_as_DDS_DomainParticipantFactory(factory);

    DDS_DomainParticipant dp = DDS_DomainParticipantFactory_create_participant(dds_factory, domain_id, NULL, NULL, 0);
    if (!dp) {
        fprintf(stderr, "FAIL: create_participant() failed on domain %u\n", domain_id);
        return 1;
    }

    if (zzdds_register_type_support(dp, "HistoryEvent", HistoryEvent_compute_key_hash_from_cdr, HistoryEvent_get_field_from_cdr) != DDS_RETCODE_OK) {
        fprintf(stderr, "FAIL: register_type_support failed\n");
        return 1;
    }

    DDS_Topic topic = DDS_DomainParticipant_create_topic(dp, "HistoryEvent", "HistoryEvent", NULL, NULL, 0);
    if (!topic) {
        fprintf(stderr, "FAIL: create_topic() failed\n");
        return 1;
    }
    printf("Create topic: HistoryEvent\n");

    DDS_Publisher pub = DDS_DomainParticipant_create_publisher(dp, NULL, NULL, 0);
    if (!pub) {
        fprintf(stderr, "FAIL: create_publisher() failed\n");
        return 1;
    }

    DDS_DataWriterQos dw_qos;
    DDS_Publisher_get_default_datawriter_qos(pub, &dw_qos);
    dw_qos.reliability.kind = DDS_ReliabilityQosPolicyKind_RELIABLE_RELIABILITY_QOS;
    dw_qos.durability.kind = DDS_DurabilityQosPolicyKind_TRANSIENT_LOCAL_DURABILITY_QOS;
    dw_qos.history.kind = DDS_HistoryQosPolicyKind_KEEP_ALL_HISTORY_QOS;

    DDS_DataWriter dw = DDS_Publisher_create_datawriter(pub, topic, &dw_qos, NULL, 0);
    if (!dw) {
        fprintf(stderr, "FAIL: create_datawriter() failed\n");
        return 1;
    }
    printf("Create writer for topic: HistoryEvent\n");

    PubState state;
    atomic_init(&state.ever_matched, false);
    atomic_init(&state.matched_current_count, 0);

    DDS_DataWriterListener listener;
    memset(&listener, 0, sizeof(listener));
    listener.listener_data = &state;
    listener.on_publication_matched = on_publication_matched;
    if (DDS_DataWriter_set_listener(dw, &listener, DDS_PUBLICATION_MATCHED_STATUS) != DDS_RETCODE_OK) {
        fprintf(stderr, "FAIL: set_listener failed\n");
        return 1;
    }

    HistoryEventDataWriter writer;
    HistoryEventDataWriter_init(&writer, dw, ZIDL_XCDR1);

    /* -- Historical batch: written immediately, no reader matched yet. -- */
    int seq = 0;
    for (; seq < HISTORICAL_COUNT; seq++) {
        HistoryEvent sample;
        memset(&sample, 0, sizeof(sample));
        sample.seq_num = seq;
        if (HistoryEventDataWriter_write(&writer, &sample, DDS_HANDLE_NIL) != 0) {
            fprintf(stderr, "FAIL: write() failed at seq_num=%d\n", seq);
            return 1;
        }
        printf("Publisher: wrote historical seq_num=%d\n", seq);
    }

    /* -- Wait for the late-joining reader to match. -- */
    for (int waited_ms = 0; !atomic_load(&state.ever_matched); waited_ms += POLL_PERIOD_MS) {
        if (waited_ms >= MATCH_TIMEOUT_MS) {
            fprintf(stderr, "FAIL: no reader matched within %ds\n", MATCH_TIMEOUT_MS / 1000);
            return 1;
        }
        usleep(POLL_PERIOD_MS * 1000);
    }
    printf("Publisher: reader matched, writing live batch\n");

    /* -- Live batch. -- */
    for (; seq < HISTORICAL_COUNT + LIVE_COUNT; seq++) {
        HistoryEvent sample;
        memset(&sample, 0, sizeof(sample));
        sample.seq_num = seq;
        if (HistoryEventDataWriter_write(&writer, &sample, DDS_HANDLE_NIL) != 0) {
            fprintf(stderr, "FAIL: write() failed at seq_num=%d\n", seq);
            return 1;
        }
        printf("Publisher: wrote live seq_num=%d\n", seq);
    }

    for (int waited_ms = 0;
         !(atomic_load(&state.ever_matched) && atomic_load(&state.matched_current_count) == 0);
         waited_ms += POLL_PERIOD_MS) {
        if (waited_ms >= DRAIN_TIMEOUT_MS) {
            fprintf(stderr, "FAIL: subscriber did not disconnect within %ds\n", DRAIN_TIMEOUT_MS / 1000);
            return 1;
        }
        usleep(POLL_PERIOD_MS * 1000);
    }

    printf("Publisher: done.\n");
    zzdds_destroy_factory(factory);
    return 0;
}
